### PR TITLE
Added support for features in surface layers

### PR DESCRIPTION
**Pull request by jo-mueller**
# References and relevant issues
Partial fix of #5205 

# Description
This PR adds the most basic form of support for adding the `features` to the `Surface` layer. In essence, I copied the respective functionality from the `Points` layer into the code for the Surface layer. *Note*: As discussed in the [above issue](#5205) this currently modifies the Surface layer only in so far as it does not throw an error if a `features` keyword is passed to the instantiation, but it does not hook up the coloring scheme to the features table.

### CODE REVIEW

File: napari/layers/surface/surface.py
Comment (andy-sweet) [inline napari/layers/surface/surface.py]: My initial thought was to omit the `properties` and `property_choices` parameters here. Do you need them for your work? Or would `features` and `feature_defaults` be enough?

Like you mentioned, there's the consistency to worry about (where the other layers with features do have `properties` and `property_choices`), but I wouldn't worry about that too much yet (since the `Surface` layer doesn't have either of these). And I prefer having less than more for now, especially when it comes to napari's public API and especially because I'm hoping to deprecate `properties` and `property_choices` in #5477.

If we have these parameters, then we might need the equivalent `properties` `@property`, which feels silly if we're just going to deprecate and remove them imminently.

What do you think? If you do remove these, then you should also remove the docstrings and events you initially added.

File: napari/layers/surface/surface.py
Comment (andy-sweet) [inline napari/layers/surface/surface.py]: The `Surface` layer doesn't have text (unlike `Points` and `Shapes`), so I think we can just remove this for now.

```suggestion
```

That should fix the current typing complaint on CI.

We could add `Surface.text` (or similar) in the future, but that's definitely another PR!

File: napari/layers/surface/surface.py
Comment (andy-sweet) [inline napari/layers/surface/surface.py:466]: You'll need to manually add a `features` event in `self.events.add` in `__init__`.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Looks roughly like what I was expecting - please don't take the multiple comments (or the red request changes) on the small PR as a bad sign!

My main requests here are to make the API addition smaller (by not introducing `properties`) and to add a few non-obvious missing things needed when adding a new parameter/attribute in napari.

One missing thing here I think you'll need to make the tests pass is to add `features` and `feature_defaults` to `Surface._get_state` in a similar way to the `Points._get_state`: https://github.com/napari/napari/blob/1779bd50cfacdf2bbc195e9ca8e31248c82e6c51/napari/layers/points/points.py#L1351

File: napari/layers/surface/surface.py
Comment (jo-mueller) [reply napari/layers/surface/surface.py]: Thanks for the review. I totally agree, I just put `.properties` there for consistency, while knowing that they are supposed to be deprecated in the near future. I don't rely on `.properties` anywhere in my work so I can totally remove them again.

**Comment by melonora** [comment]
Hi @jo-mueller, Thanks for your first contribution and welcome to napari! I see @andy-sweet already had a look. I will also do so later. Don't worry about us not going easy, we are a friendly bunch:)

File: napari/layers/surface/surface.py
Comment (jo-mueller) [reply napari/layers/surface/surface.py]: Removed `.properties` again in faa9f44ea6fc4553a2f265a013c653385a44035f and 4497d1b34ac582bb58b54427ff5ba6418f5419be

File: napari/layers/surface/surface.py
Comment (jo-mueller) [reply napari/layers/surface/surface.py:466]: Added in 13119314cb4bbc27a4d1f7274448d1d481706fdd

**Comment by andy-sweet** [comment]
> One missing thing here I think you'll need to make the tests pass is to add `features` and `feature_defaults` to `Surface._get_state` in a similar way to the `Points._get_state`:

@jo-mueller : no pressure or rush, but I made a PR on top of this one to suggest the associated code changes I was thinking about that should make the tests pass and maybe have this PR in good shape to be merged. If it looks good to you (and the tests actually pass!), feel free to merge that into your branch or just copy the code.

**Comment by jo-mueller** [comment]
@andy-sweet Thanks for the contribution, it's nice to cooperate this way 👍

File: napari/layers/surface/surface.py
Comment (brisvag) [inline napari/layers/surface/surface.py]:
```suggestion
        Features table where each row corresponds to a vertex and each column
        is a feature.
```

**Review by brisvag — APPROVED** [review]
Nice work @jo-mueller! Straight forward PR to review :) Are you planning on following up with some more logic for coloring/selecting and so on? Looking forward to the discussion there!

**Comment by jo-mueller** [comment]
Hi @brisvag ,

thanks for the review! We discussed (see #5205 ) to maybe add some functionality so that a surface layer wouldn't throw an error anymore if `features` were passed to it (this PR). As a next step we were thinking about hooking up the vertex coloring to the features and the last step somewhere down the road would be to deprecate the `values` in the `(vertices, faces, values)` surface tuple schema.

So...I would try myself at wiring the color functionality up in a next PR and I'd definitely happy for any pointers in the right direction :)

**Comment by andy-sweet** [comment]
@jo-mueller : almost ready to approve, but noticed a bug in the setter.

Fix suggestion with test in https://github.com/jo-mueller/napari/pull/2 

Again, feel free to merge, copy those changes in, or make an alternate suggestion.

**Review by andy-sweet — APPROVED** [review]
Looks good now - thanks!

**Comment by andy-sweet** [comment]
@jo-mueller : I updated the PR description to be more concise and fixed a typo, as we use that as the commit message.

@Czaki : should I wait for the v0.4.19 release before adding the `ready to merge` label or merging this? The changes here are pretty isolated, but we can also wait a little.

**Comment by andy-sweet** [comment]
@Czaki : is it safe to merge this yet? i noticed some other PRs targeted for 0.5.0 were merged recently

**Comment by Czaki** [comment]
I think yes. Waiting PR do not touch this files, and I do not get found any bug reports on 0.4.19 touching surface

**Comment by andy-sweet** [comment]
Thanks! I'll merge this at the end of my tomorrow, unless anyone objects before then.

**Comment by Czaki** [comment]
add label

**Pull request merged**
